File: hooks/repo_hygiene_check.go
package main

import (
	"fmt"
	"os"
)

// RequiredCandidate represents a file or directory that must exist with a specific type.
type RequiredCandidate struct {
	Path         string // Relative path from the repository root.
	RequiredType string // Expected type: "file" or "dir".
}

// List of required files and directories.
var requiredCandidates = []RequiredCandidate{
	{Path: ".github", RequiredType: "dir"},
	{Path: ".github/CODEOWNERS", RequiredType: "file"},
	{Path: ".github/PULL_REQUEST_TEMPLATE.md", RequiredType: "file"},
	{Path: ".harness", RequiredType: "dir"},
	{Path: ".harness/piplines", RequiredType: "dir"},
	{Path: ".harness/input_steps", RequiredType: "dir"},
	{Path: ".vscode", RequiredType: "dir"},
	{Path: ".vscode/extentions.json", RequiredType: "file"},
	{Path: ".dockerignore", RequiredType: "file"},
	{Path: ".editorconfig", RequiredType: "file"},
	{Path: ".gitignore", RequiredType: "file"},
	{Path: ".pre-commit-config.yaml", RequiredType: "file"},
}

func main() {
	missingFound := false

	// Loop over each required entry, checking for both existence and type.
	for _, candidate := range requiredCandidates {
		info, err := os.Stat(candidate.Path)
		if err != nil {
			fmt.Printf("WARNING: Missing required %s: %s\n", candidate.RequiredType, candidate.Path)
			missingFound = true
			continue
		}

		// Verify that the candidate is of the required type.
		if candidate.RequiredType == "file" && info.IsDir() {
			fmt.Printf("WARNING: Expected file but found directory: %s\n", candidate.Path)
			missingFound = true
		} else if candidate.RequiredType == "dir" && !info.IsDir() {
			fmt.Printf("WARNING: Expected directory but found file: %s\n", candidate.Path)
			missingFound = true
		}
	}

	// Print an overall summary.
	if missingFound {
		fmt.Println("⚠️ Repository hygiene check: some required files/directories are missing or incorrect.")
	} else {
		fmt.Println("✅ Repository hygiene check passed.")
	}

	// Always exit with 0 so that the commit is not blocked.
	os.Exit(0)
}

File: hooks/repo_hygiene_check.py
#!/usr/bin/env python3
import os
import sys

def main():
    # Print the current working directory for debugging purposes.
    work_dir = os.getcwd()
    print(f"Running repository hygiene check in: {work_dir}")

    # Define a list of required items with their expected type.
    # Each item is a dictionary with:
    #   "path": the relative path to check
    #   "type": either "file" or "dir"
    required_items = [
        {"path": ".github", "type": "dir"},
        {"path": ".github/CODEOWNERS", "type": "file"},
        {"path": ".github/PULL_REQUEST_TEMPLATE.md", "type": "file"},
        {"path": ".harness", "type": "dir"},
        {"path": ".harness/piplines", "type": "dir"},
        {"path": ".harness/input_steps", "type": "dir"},
        {"path": ".vscode", "type": "dir"},
        {"path": ".vscode/extentions.json", "type": "file"},
        {"path": ".dockerignore", "type": "file"},
        {"path": ".editorconfig", "type": "file"},
        {"path": ".gitignore", "type": "file"},
        {"path": ".pre-commit-config.yaml", "type": "file"},
    ]

    missing_found = False

    # Check each required path.
    for item in required_items:
        path = item["path"]
        expected_type = item["type"]

        if not os.path.exists(path):
            # The candidate does not exist at all.
            print(f"WARNING: Missing required {expected_type}: {path}")
            missing_found = True
        else:
            if expected_type == "file" and not os.path.isfile(path):
                # Path exists but is not a file as expected.
                print(f"WARNING: Expected a file but found something else: {path}")
                missing_found = True
            elif expected_type == "dir" and not os.path.isdir(path):
                # Path exists but is not a directory as expected.
                print(f"WARNING: Expected a directory but found something else: {path}")
                missing_found = True

    # Overall summary: print whether the hygiene check passed or if some items are missing.
    if missing_found:
        print("⚠️  Repository hygiene check: some required files/directories are missing or incorrect.")
    else:
        print("✅ Repository hygiene check passed.")

    # Exit with code 0 so that the commit is not blocked.
    sys.exit(0)

if __name__ == "__main__":
    main()
